Initial code structure for Andersen Alias analysis

haved opened the pull request:
It contains minimal optimizations for constraint solving, but tries to avoid excessive heap usage.

It might be better to move more of the struct definitions into headers, to make it possible to write unit tests for the set constraint solving itself, without an RVSDG graph in between.

There are also many statistics that it would make sense to track, such as the number of each constraint, how long solving takes, how many iterations and how long building the PtG takes.

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+   * Internal helper function for adding nodes, use the Create* methods instead
+   */
+  PointerObjectIndex
+  AddNode(PointerObjectKind kind)
```
`AddPointerObject()`?

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+
+public:
+  PointerObject &
+  GetNode(PointerObjectIndex index)
```
`GetPointerObject()`?

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+  }
+
+  PointerObjectIndex
+  CreateRegisterNode(jlm::rvsdg::output *rvsdgOutput)
```
`CreateRegisterNode(rvsdg::output & rsvdgOutput)`?

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+  }
+
+  void
+  MapRegisterToExistingNode(jlm::rvsdg::output *rvsdgOutput, PointerObjectIndex node)
```
`MapRegisterToExistingNode(rvsdg::output & rvsdgOutput, ....)`

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+  }
+
+  PointerObjectIndex
+  CreateAllocaMemoryObject(jlm::rvsdg::node *allocaNode)
```
`CreateAllocaMemoryObject(rvsdg::node & allocaNode)`

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+  }
+
+  PointerObjectIndex
+  CreateMallocMemoryObject(jlm::rvsdg::node *mallocNode)
```
`CreateMallocMemoryObject(rvsdg::node & mallocNode)`

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+  }
+
+  PointerObjectIndex
+  CreateGlobalMemoryObject(jlm::llvm::delta::node *deltaNode)
```
`CreateGlobalMemoryObject(delta::node & deltaNode)`

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+  }
+
+  PointerObjectIndex
+  CreateFunction(jlm::llvm::lambda::node *lambdaNode)
```
`CreateFunction(lambda::node & lambdaNode)`

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+  }
+
+  PointerObjectIndex
+  CreateImportMemoryObject(jlm::rvsdg::argument *importNode)
```
`CreateImportMemoryObject(rvsdg::argument & importNode)`

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+  {
+    auto pointsToGraph = PointsToGraph::Create();
+
+    // memory nodes are the nodes that can be pointed to in the PtG.
```
`PtG` spell it out to `points-to graph`. I do not think this abbreviation is used somewhere else in the code.

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+      memoryNodes[pointerObjectIndex] = &node;
+    }
+
+    auto applyPointsToSet = [&](PointsToGraph::Node* node, PointerObjectIndex index) {
```
`(PointsGraph::Node & node, ....)`

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+  PointerObjectIndex Pointer1_;
+  PointerObjectIndex Pointer2_;
+
+public:
+  AllPointeesPointToSupersetConstraint(PointerObjectIndex pointer1, PointerObjectIndex pointer2)
+          : Pointer1_(pointer1), Pointer2_(pointer2) {}
+
```
Indentation

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+ * for all x in P(pointer1), make P(x) a superset of P(pointer2)
+ * Example of application is a store, e.g. when *pointer1 = pointer2
+ */
+class AllPointeesPointToSupersetConstraint final {
```
Curly braces on new line, please

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+ * P(loaded) = union of P(x) for all x in P(pointer)
+ * Example of application is a load, e.g. when loaded = *pointer
+ */
+class SupersetOfAllPointeesConstraint final {
```
Curly braces on new line, please

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+  PointerObjectIndex Pointer_;
+
+public:
+  SupersetOfAllPointeesConstraint(PointerObjectIndex loaded, PointerObjectIndex pointer)
+          : Loaded_(loaded), Pointer_(pointer) {}
```
Indentation?

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+  }
+};
+
+using ConstraintVariant = std::variant<AllPointeesPointToSupersetConstraint, SupersetOfAllPointeesConstraint>;
```
Move it in the PionterObjectConstraintSet class?

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+  std::vector<ConstraintVariant> Constraints_;
+};
+
+void
+PointerObjectConstraintSet::Solve()
+{
+  // Keep applying constraints until no sets are modified
+  bool modified = true;
+  while (modified) {
+    modified = false;
+
+    for (auto& constraint : Constraints_)
+      std::visit([&](auto constraint) {
+        modified |= constraint.Apply(Set_);
+      }, constraint);
+
+    modified |= ApplyEscapedFlagPropagation();
+  }
+}
+
+bool
+PointerObjectConstraintSet::ApplyEscapedFlagPropagation()
+{
+  // This function only applies one iteration of escape flag propagation, so it might need to be run again
+  bool modified = false;
+
+  for (PointerObjectIndex idx = 0; idx < Set_.NodeCount(); idx++)
+  {
+    if (!Set_.GetNode(idx).HasEscaped())
+      continue;
+
+    for (PointerObjectIndex pointee : Set_.GetPointsToSet(idx)) {
+      auto& pointeeNode = Set_.GetNode(pointee);
+      modified |= !pointeeNode.HasEscaped();
+      pointeeNode.MarkAsEscaped();
+    }
+  }
+
+  return modified;
+}
```
Why not defined inlined?

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+std::unique_ptr<PointsToGraph>
+Andersen::Analyze(const RvsdgModule &module, jlm::util::StatisticsCollector &statisticsCollector)
+{
+  Set_ = std::make_unique<PointerObjectSet>();
+  Constraints_ = std::make_unique<PointerObjectConstraintSet>(*Set_);
```
Please call `reset()` on the `Set_` and `Constraints_` unique pointers before you exit the function.

phate commented on jlm/llvm/opt/alias-analyses/Andersen.hpp:
```diff
@@ -0,0 +1,120 @@
+#include <jlm/llvm/ir/operators.hpp>
+#include <jlm/llvm/opt/alias-analyses/AliasAnalysis.hpp>
+
+namespace jlm::llvm::aa {
```
Curly braces on new line, please.

phate commented on jlm/llvm/opt/alias-analyses/Andersen.hpp:
```diff
@@ -0,0 +1,120 @@
+#ifndef JLM_LLVM_OPT_ALIAS_ANALYSES_ANDERSEN_HPP
+#define JLM_LLVM_OPT_ALIAS_ANALYSES_ANDERSEN_HPP
+
+#include <jlm/llvm/ir/RvsdgModule.hpp>
+#include <jlm/llvm/ir/operators.hpp>
```
Alphabetical order?

phate commented on jlm/llvm/opt/alias-analyses/Andersen.hpp:
```diff
@@ -0,0 +1,120 @@
+#ifndef JLM_LLVM_OPT_ALIAS_ANALYSES_ANDERSEN_HPP
+#define JLM_LLVM_OPT_ALIAS_ANALYSES_ANDERSEN_HPP
```
The header comment is missing.

phate commented on jlm/llvm/opt/alias-analyses/Andersen.hpp:
```diff
@@ -0,0 +1,120 @@
+  Andersen &
+  operator=(Andersen &&) = delete;
+
+  std::unique_ptr<PointsToGraph>
+	Analyze(
+    const RvsdgModule & module,
+    jlm::util::StatisticsCollector & statisticsCollector) override;
```
Indentation?

phate commented on jlm/util/math.hpp:
```diff
@@ -0,0 +1,35 @@
+#include <type_traits>
+
+namespace jlm::util
+{
+
+/**
+ * Log2 for integers, rounding up when value is not a perfect power of two
+ */
+template<class T>
+static inline constexpr int Log2i(T value) {
+  if (value <= 1)
+    return 0;
+  return 1 + Log2i(value >> 1);
+}
+
+/**
+ * Calculates the number of bits needed to hold the integer representation of all enum values.
+ * Takes a sentinel end value as parameter, requires all other enum values to have a lower integer value.
+ */
+template<class T>
+static inline constexpr int BitWidthOfEnum(T sentinelEndValue) {
+  static_assert(std::is_enum<T>::value, "BitWidthOfEnum only takes enums");
+
+  using UnderlyingT = std::underlying_type_t<T>;
+  return Log2i(static_cast<UnderlyingT>(sentinelEndValue));
+}
+
```
I would like to see some unit  tests on those, please.

haved replied:
You have a very good point, because that Log2i is absolutely not rounding up :upside_down_face:

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+   * @return
+   */
+  std::unique_ptr<PointsToGraph>
+  ConstructPointsToGraph()
```
The points-to graph construction method as part of the `PointerObjectSet` class feels "wrong". Should it not rather be a static private method of the Andersen class and you hand the set in as parameter? This would require that you define access methods to the private attributes though....

phate commented on jlm/util/math.hpp:
```diff
@@ -0,0 +1,35 @@
+
+
+#ifndef JLM_UTIL_MATH_HPP
```
Two new lines and no header comment.

phate commented on jlm/llvm/opt/alias-analyses/Andersen.cpp:
```diff
@@ -0,0 +1,514 @@
+public:
+  explicit PointerObjectConstraintSet(PointerObjectSet& set) : Set_(set) {}
+
+  PointerObjectConstraintSet(const PointerObjectConstraintSet& other) = delete;
+
+  PointerObjectConstraintSet&
+  operator =(const PointerObjectConstraintSet& other) = delete;
```
What about deleting move constructor and move assignment?

phate replied:
I did not know that. Learned something. Thanks! Personally, I prefer to have it explicitly instead of remembering such rules.

phate reviewed (commented):
Please do not misunderstand the number of comments as an indication about the code quality. The code is very readable and I like it a lot. Great work! I just went all out on it and commented on everything minor that caught my eye. Some of it is certainly a matter of personal preference.

I would move the PointerObject stuff into its own file though. This way you can easily unit test it.

haved commented:
Thank you for the great reviewing, your suggestions all make prefect sense, so I'll apply them, make a separate file for PointerObject stuff, and write some tests. I have also changed references to "Node" to use "PointerObject" instead.

I'm wondering what your preferred way of referring to symbols from "cousin" namespaces is. If I'm in the `jlm::llvm::aa` namespace, and want to access `jlm::llvm::delta::node`, do you prefer just `delta::node`? What about accessing `jlm::rvsdg::output`, do you include the `jlm` part? In e.g. `PointsToGraph.hpp` it seems the jlm is more often included in the second case. 

I will remove the jlm part for both, since your review indicated a preference for `rvsdg::output` etc.

The use of `std::set` was a mistake, I thought ordered sets had faster union operations, but that does not seem to be the case. If all the `P[x].insert(P[y].begin(), P[y].end())` calls end up taking a lot of time, it is probably better to implement some "dirty" tracking, and only make the calls when sets have changed. (A much cooler optimization would be to let each P-set be generational, with newly added members in a separate set until they have been new for long enough that all constraints have seen them). Then again it might be the case that 99% of sets are tiny, so I'll not implement any such things for now.

I'm a bit surprised that
```
PointerObjectKind Kind_ : jlm::util::BitWidthOfEnum(PointerObjectKind::COUNT);
```
Is a legal field, but it saves us from doing `static_cast< ... >()`, so I'm glad it is.

phate commented:
> I'm wondering what your preferred way of referring to symbols from "cousin" namespaces is. If I'm in the jlm::llvm::aa namespace, and want to access jlm::llvm::delta::node, do you prefer just delta::node? What about accessing jlm::rvsdg::output, do you include the jlm part? In e.g. PointsToGraph.hpp it seems the jlm is more often included in the second case.

As short as possible. The `jlm::` part is mostly there because of something automatic namespace refactoring.

haved commented:
 @phate I tried adding two new classes to `TestRVSDG.cpp` to fully test the `PointerObjectSet.cpp`. Please comment if you want anything different. Also, I didn't find a clean way of ensuring subclasses of `TestRVSDG` call `SetupModule()`, other than calling `module()` and discarding the result, so I added a method. If you have some other suggestion, please share.

phate commented on tests/TestRvsdgs.hpp:
```diff
@@ -18,9 +18,7 @@ class RvsdgTest {
 	jlm::llvm::RvsdgModule &
```
Let's fix the indentation for the entire function, please.

haved replied:
I changed all tabs to spaces in the RvsdgTest class, but left the rest of the file alone. Is that what you meant? I hadn't noticed before that this file contains both tabs and spaces, but it seems recent code prefers spaces?

phate replied:
Yes, we switched from tabs to spaces at some point. Now, we have a happy mix. I think it is time to get a clang format going and enforce such behavior at some point.

phate commented on tests/TestRvsdgs.hpp:
```diff
@@ -30,6 +28,13 @@ class RvsdgTest {
 	}
 
+  void
+  EnsureInitialized()
```
Could we call this function `Initialize()` (or something else) and put some documentation above it that it always needs to be called to setup the test.

phate commented on tests/TestRvsdgs.hpp:
```diff
@@ -1807,4 +1812,146 @@ class LinkedListTest final : public RvsdgTest
+ * The class sets up an RVSDG module corresponding to the code:
+ *
+ * \code{.c}
+ *   static int global = 1;
```
This is static and nowhere used in the code, and therefore the delta is not exported. This means DNE will just remove it if it is ever used on it. I am not sure that is what you want!?

haved replied:
This TestRVSDG subclass doesn't really represent a useful program, so I agree that any DNE pass applied to it would remove unused things (same applies to the import and alloca). I can try to make the program a bit more realistic, to perhaps make the test class more generally useful.

phate replied:
It is fine as it is, I just wanted to point it out in  case it was not a wanted side-effect.

phate replied:
Btw, did you make any changes after my last round of comments? If so, did you forget to push them as I see no differences?

phate commented on tests/TestRvsdgs.hpp:
```diff
@@ -1807,4 +1812,146 @@ class LinkedListTest final : public RvsdgTest
+  std::unique_ptr<jlm::llvm::RvsdgModule>
+  SetupRvsdg() override;
+
+  size_t NumAllocaNodes_;
```
You might want to add a Getter for the `NumAllocaNodes_` as well.

phate commented on tests/TestRvsdgs.hpp:
```diff
@@ -1807,4 +1812,146 @@ class LinkedListTest final : public RvsdgTest
+class NAllocaNodesTest final : public RvsdgTest
+{
+public:
+  NAllocaNodesTest(size_t numAllocaNodes) : NumAllocaNodes_(numAllocaNodes) {}
```
```
NAllocNodesTest(size_t numAllocaNodes)
  : NumAllocaNodes_(numAllocaNodes)
{}
```
It is more readable, one piece of information per line.

phate commented:
> @phate I tried adding two new classes to `TestRVSDG.cpp` to fully test the `PointerObjectSet.cpp`. Please comment if you want anything different. Also, I didn't find a clean way of ensuring subclasses of `TestRVSDG` call `SetupModule()`, other than calling `module()` and discarding the result, so I added a method. If you have some other suggestion, please share.

I only looked through the TestRvsdg.hpp/*.cpp files. I sadly do not have another suggestion. I was aware of this shortcoming in the TestRvsdg class, but had no better idea either. The introduction of your method would make it the initialization at least explicit and add some documentation. I like that.

haved commented:
@phate I removed the draft tag on this PR, as it should be ready to be merged as a standalone unit.
From last time you looked at it:
 - I removed the Andersen.[ch]pp files from this PR (including the PointerObjectSet-> PointsToGraph conversion code)
 - Fixed the `AllMemoryNodesTest` graph a bit, to not have any obviously dead nodes.
 - Added tests of the methods on `PointerObjectConstraintSet`.
 - Added the implication Escaped -> PointsToExternal, as we established at the meeting.

Each constraint was already tested on its own, but to test the constraint solver itself, I had to combine multiple constraints on a somewhat larger graph. The result is some long and dense testing functions, with `TestPointerObjectConstraintSetSolve()` trying to use pretty much every type of constraint.

I hope that no more constraints will have to be added to handle all of Andersen, so it should hopefully stay quite static from here on out.

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp:
```diff
@@ -0,0 +1,369 @@
+  GetPointsToSet(PointerObject::Index idx) const;
+
+  /**
+   * Adds pointee to P(pointer)
```
You have nowhere defined what `P(pointer)` is. Maybe: "Adds \p pointee to pointer object set of \p pointer"

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp:
```diff
@@ -0,0 +1,369 @@
+
+  /**
+   * Adds pointee to P(pointer)
+   * @param pointer the index of the PointerObject that points
```
"...that points" where to? The sentence is incomplete

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp:
```diff
@@ -0,0 +1,369 @@
+  AddToPointsToSet(PointerObject::Index pointer, PointerObject::Index pointee);
+
+  /**
+   * Makes P(superset) a superset of P(subset) by adding any elements in the set difference
```
Similarly here, P(...) was nowhere defined.

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp:
```diff
@@ -0,0 +1,369 @@
+  MakePointsToSetSuperset(PointerObject::Index superset, PointerObject::Index subset);
+
+  /**
+   * Adds the Escaped flag to all PointerObjects in the P(pointer) set
```
The same here.

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp:
```diff
@@ -0,0 +1,369 @@
+  PointerObject::Index Subset_;
+
+public:
+  SupersetConstraint(PointerObject::Index superset, PointerObject::Index subset)
+          : Superset_(superset), Subset_(subset) {}
```
Indentation, one at a time, please.

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp:
```diff
@@ -0,0 +1,369 @@
+  MarkAllPointeesAsEscaped(PointerObject::Index pointer);
+};
+
+/**
+ * A constraint of the form:
+ * P(superset) is a superset of P(subset)
+ * Example of application is when a register has multiple source values
+ */
+class SupersetConstraint final
```
Okay, by now I would suggest to you rather introduce P(...) somewhere instead of rewriting comments.

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp:
```diff
@@ -0,0 +1,369 @@
+          : Superset_(superset), Subset_(subset) {}
+
+  bool
+  Apply(PointerObjectSet& set);
```
What does this function do? Apply is a rather generic name? Some documentation might help.

phate replied:
Meh, after reading it a second time, it obviously applies the constraint. Maybe still some documentation that highlights this for sleepy readers....

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp:
```diff
@@ -0,0 +1,369 @@
+  PointerObject::Index Pointer2_;
+
+public:
+  AllPointeesPointToSupersetConstraint(PointerObject::Index pointer1, PointerObject::Index pointer2)
+                                      : Pointer1_(pointer1), Pointer2_(pointer2) {}
+
```
Indentation, one at a time, please.

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp:
```diff
@@ -0,0 +1,369 @@
+                                      : Pointer1_(pointer1), Pointer2_(pointer2) {}
+
+  bool
+  Apply(PointerObjectSet& set);
```
Same here. very generic name. Documentation?

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp:
```diff
@@ -0,0 +1,369 @@
+  PointerObject::Index Pointer_;
+
+public:
+  SupersetOfAllPointeesConstraint(PointerObject::Index loaded, PointerObject::Index pointer)
+                                 : Loaded_(loaded), Pointer_(pointer) {}
```
Indentation, one at a time.

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp:
```diff
@@ -0,0 +1,369 @@
+                                 : Loaded_(loaded), Pointer_(pointer) {}
+
+  bool
+  Apply(PointerObjectSet& set);
```
Same here

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp:
```diff
@@ -0,0 +1,369 @@
+ * Unlike the set modification methods on PointerObjectSet, constraints can be added in any order, with the same result.
+ * Use Solve() to calculate the final points-to-sets.
+ *
+ * Constraints on the special nodes, external and escaped, are built in.
```
External and Escaped are special nodes? What nodes?

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp:
```diff
@@ -0,0 +1,369 @@
+          AllPointeesPointToSupersetConstraint,
+          SupersetOfAllPointeesConstraint>;
+
+  explicit PointerObjectConstraintSet(PointerObjectSet& set) : Set_(set) {}
```
Indentation

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp:
```diff
@@ -0,0 +1,369 @@
+  AddConstraint(ConstraintVariant c);
+
+  /**
+   * Iterates over and applies constraints until the all points-to-sets satisfy them
```
"...until the all..." -> "...until all..."

This function sounds "expensive". Can you say anything about its runtime performance?

haved replied:
It most definitely is expensive, as in O(n^3). In practice it should be a bit better if the constraint order is mostly the same as data flow order.

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp:
```diff
@@ -0,0 +1,369 @@
+private:
+
+  /**
+   * Makes sure any PointerObject marked as escaped, also makes all its pointees get the HasEscaped flag set.
```
"Ensures that the escaped flag is set for all pointees of any pointer object that is marked as escaped, "

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp:
```diff
@@ -0,0 +1,303 @@
+}
+
+void
+PointerObjectSet::MapRegisterToExistingPointerObject(const rvsdg::output & rvsdgOutput, PointerObject::Index pointerObject)
```
each argument on new line, please

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp:
```diff
@@ -0,0 +1,369 @@
+  const std::unordered_map<const rvsdg::argument *, PointerObject::Index>&
+  GetImportMap() const noexcept;
+
+  PointerObject::Index
```
I would have expected a `size_t` as return type, not a `PointerObject::Index`

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp:
```diff
@@ -0,0 +1,303 @@
+  // Registers can not be pointed to
+  JLM_ASSERT(GetPointerObject(pointee).GetKind() != PointerObjectKind::Register);
+
+  auto sizeBefore = PointsToSets_[pointer].size();
+  PointsToSets_[pointer].insert(pointee);
+
+  return PointsToSets_[pointer].size() != sizeBefore;
```
Does the[ insert() method from std::unordered_set ](https://en.cppreference.com/w/cpp/container/unordered_set/insert) not already give you that information?

Could you not replace this code with:

```
return PointsToSets_[pioniter].insert(pointee).second
```

I am not saying it is more readable....

Maybe you want to consider to use [HashSet](https://github.com/phate/jlm/blob/master/jlm/util/HashSet.hpp)?

haved replied:
`std::unordered_set::insert()` absolutely does, I will use that.

I can use HashSet, I didn't know in what ways it differed from unordered_set, so I went with what I already (kind of) knew.

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp:
```diff
@@ -0,0 +1,303 @@
+    modified |= set.MakePointsToSetSuperset(Loaded_, x);
+  }
+
+  // Handling pointing to external is done by MakePointsToSetSuperset,
+  // Propagating escaped status is handled by different constraints
```
Thanks for the comments.

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp:
```diff
@@ -0,0 +1,303 @@
+
+
+void
+PointerObjectConstraintSet::AddPointerPointeeConstraint(PointerObject::Index pointer, PointerObject::Index pointee)
```
Both on new line, please.

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp:
```diff
@@ -0,0 +1,303 @@
+}
+
+void
+PointerObjectConstraintSet::Solve() {
```
Parenthesis on new line, please.

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp:
```diff
@@ -0,0 +1,303 @@
+  while (modified) {
+    modified = false;
+
+    for (auto& constraint : Constraints_)
+      std::visit([&](auto constraint) {
+        modified |= constraint.Apply(Set_);
+      }, constraint);
```
```
for (auto& constraint : Constraints_)
{
  std::visit(
    [&](auto  & constraint) { modified |= constraint.Apply(Set_); },
    constraint);
}
```
Please mind the `auto & constraint` instead of the `auto constraint` from the original code. None of those constraints has a deleted copy constructor. So not using an `&` should copy this stuff happily around, no? Is this what you want here?

haved replied:
I didn't intend to use a reference only for the loop and not the call. I know most constraints are just two 64-byte words, so they should fit in registers, but the call to Apply passes `this` as a pointer, so it doesn't matter. If we are lucky, the entire call to Apply gets inlined.

phate commented on jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp:
```diff
@@ -0,0 +1,303 @@
+bool
+PointerObjectConstraintSet::PropagateEscapedFlag()
+{
+  bool modified = false;
```
Could we move this line of code between the while and for loop. It is not used in the first part of the code, i.e., the while loop.

phate commented on tests/jlm/llvm/opt/alias-analyses/TestPointerObjectSet.cpp:
```diff
@@ -0,0 +1,516 @@
+  return 0;
+}
+
+JLM_UNIT_TEST_REGISTER(
+"jlm/llvm/opt/alias-analyses/TestPointerObjectSet",
+TestPointerObjectSet)
```
Indent, please.

phate commented on tests/jlm/llvm/opt/alias-analyses/TestPointerObjectSet.cpp:
```diff
@@ -0,0 +1,516 @@
+{
+  using namespace jlm::llvm::aa;
+
+  /*
+   * Line in the test corresponds to lines of pseudo-SSA, with branching removed.
+   * First, the function definition and local variables.
+   *
+   * function f(%0):
+   * %1 = alloca 8 (variable v1)
+   * %2 = alloca 8 (variable v2)
+   * %3 = alloca 8 (variable v3)
+   * %4 = alloca 8 (variable v4)
+   */
+
```
Could we use `//` instead of `/* ... */`

phate replied:
I do not fully understand what this comment means?

phate reviewed (changes requested):
This was a pleasure to read. Thanks a lot. Great work!

I am really looking forward to the rest of the analysis code. I am so excited to get this in jlm. 

I went all out on commenting again. Most of it is absolutely minor stuff. I would still like for you to fix at least the 1-2 more "serious" comments.

haved commented:
I think I fixed all your review comments, I'm compiling my "responses" here:

I can replace `std::unordered_set` with `HashSet`, if that is preferred. At the moment I have no idea when you would prefer one or the other. Do they have different features, or different performance or memory characteristics?

`PointerObjectConstraintSet::Solve()` can absolutely be slow. In the literature, Andersen is frequently referred to as a O(n^3) algorithm. The current solver is not at all trying to be clever, just blindly applying constraints, which can lead to the following performance pitfalls:
  - A chain of superset constraints of length N, if added in the opposite order of the "set membership flow", will need N full iterations though all the constraints, to pass a set member all the way through.
  - When applying a superset constraint, all members in the subset M will be iterated over and attempted inserted, even if they are already in the superset N. This gives the subset constraint a Theta(|M| log |N|) runtime, even when no actual "progress" is being made.

I have some thoughts about improving the performance, such as creating a graph of constraints and trying to *mostly* topologically sort it, and keeping a set membership history to only process set members that have been added since the last time the constraint was processed. I have also found several papers discussing optimizations, which should help. Before any of that, though, I want to get the baseline in.

Finally, regarding the use of `[&](auto & constraint)`, I think passing by reference makes the most sense, since the Apply call takes `this` by pointer. We can also hope everything gets inlined through the `std::visit`, in which case it doesn't matter.

phate commented:
> I can replace std::unordered_set with HashSet, if that is preferred. At the moment I have no idea when you would prefer one or the other. Do they have different features, or different performance or memory characteristics?

I prefer HashSet. HashSet is nothing else than a wrapper around std::unordered_set with a sane interface. I got tired of sometimes deciphering the interface from std::unordered_set. It is not very readable. Best example is the change I suggested yesterday: `....Insert(...).second`.

> PointerObjectConstraintSet::Solve() can absolutely be slow. In the literature, Andersen is frequently referred to as a O(n^3) algorithm.

I was aware of the O(n^3) runtime. The point of my comment was to maybe add this piece of information to the documentation. I should have written it more clearly.

> I have some thoughts about improving the performance, such as creating a graph of constraints and trying to mostly topologically sort it, and keeping a set membership history to only process set members that have been added since the last time the constraint was processed. I have also found several papers discussing optimizations, which should help. Before any of that, though, I want to get the baseline in.

You just gave a TLDR of your master thesis. The seven papers I sent you try to do exactly this: decrease the constant of this algorithm such it becomes usable for practical programs.

> Before any of that, though, I want to get the baseline in.

Yes, please. This is the most important part now.

> Finally, regarding the use of [&](auto & constraint), I think passing by reference makes the most sense, since the Apply call takes this by pointer. We can also hope everything gets inlined through the std::visit, in which case it doesn't matter.

I see no benefit in passing by value here. Regarding inlining: So I had to look this up, but as I understand it, the compiler will most likely inline std::visit, which under the hood will be nothing else than a switch statement on the underlying type(s), and then depending on the type, the right `Apply` will be called, which might be inlined as well. So if I understand this correctly, then this is nothing else than a dispatch table that is created for a virtual method. Is this correct? Why did you go this way instead of using a class with an abstract `Apply` method?

Okay, after writing the above paragraph, I googled a little bit to get some of my questions answered. It is a poor man's sum types. My next follow up question: Do we expect some other kind of constraints to pop up at some point (such that we have to change all the visitors)?

haved commented:
I hadn't realized it was a nicer interface around unordered_set, I will switch to it.

I will add documentation about the runtime of Solve().

I don't imagine any other constraints will need to be added, based on my current Andersen class, and even if it is needed, it should not affect any other code, since there is only one std::visit, and no way of extracting constraints from the ConstraintSet.

I will make these changes, rebase to get the clang-format, then I'll be very happy to get this merged.

haved commented:
I did what I described above, feel free to merge once the tests have passed. I only applied clang-format to files added by this PR.

I also added a bool return to HashSet's `UnionWith` returning true if modified, hope that's OK.

Feel free to merge once the tests have passed. I'm not sure what's going on with hls-test-suite at the moment, but it seems to affect the master branch as well.
